File: Domain/Puzzles/EverySecondTurn/tests/EverySecondTurnSolver_test.py
import unittest
from unittest import TestCase

from Domain.Board.Grid import Grid
from Domain.Board.IslandsGrid import IslandGrid
from Domain.Puzzles.EverySecondTurn.EverySecondTurnSolver import EverySecondTurnSolver

_ = '.'
X = '*'


class EverySecondTurnSolverTests(TestCase):
    def test_solution_6_6_37g81(self):
        """https://gridpuzzle.com/every-second-turn/37g81"""

        grid = Grid([
            [_, _, _, _, _, X],
            [_, X, _, _, _, _],
            [_, _, _, _, X, _],
            [X, _, _, X, _, _],
            [_, _, X, _, _, X],
            [X, _, _, _, _, _],
        ])
        expected_solution_str = (
            ' ┌──────────────┐ \n'
            ' │  ┌────────┐  │ \n'
            ' │  │  ┌─────┘  │ \n'
            ' └──┘  │  ┌─────┘ \n'
            ' ┌─────┘  └─────┐ \n'
            ' └──────────────┘ '
        )

        game_solver = EverySecondTurnSolver(grid)
        solution = game_solver.get_solution()
        self.assertEqual(expected_solution_str, str(solution))
        other_solution = game_solver.get_other_solution()
        self.assertEqual(IslandGrid.empty(), other_solution)

    def test_solution_6_6_3e9k2(self):
        """https://gridpuzzle.com/every-second-turn/3e9k2"""

        grid = Grid([
            [X, _, _, _, _, _],
            [_, _, _, _, _, X],
            [_, _, X, _, _, _],
            [_, _, _, _, X, _],
            [_, X, _, _, _, _],
            [_, _, _, _, _, X]
        ])
        expected_solution_str = (
            ' ┌──────────────┐ \n'
            ' │  ┌───────────┘ \n'
            ' │  │  ┌────────┐ \n'
            ' │  │  └─────┐  │ \n'
            ' │  └────────┘  │ \n'
            ' └──────────────┘ '
        )

        game_solver = EverySecondTurnSolver(grid)
        solution = game_solver.get_solution()
        self.assertEqual(expected_solution_str, str(solution))
        other_solution = game_solver.get_other_solution()
        self.assertEqual(IslandGrid.empty(), other_solution)

    def test_solution_10_10_0ygm8(self):
        """https://gridpuzzle.com/every-second-turn/0ygm8"""

        grid = Grid([
            [X, _, _, _, _, X, _, _, X, _],
            [_, _, _, X, _, _, _, X, _, _],
            [_, _, X, _, _, _, X, _, _, _],
            [_, X, _, _, X, _, _, _, X, _],
            [X, _, _, _, X, _, _, _, _, _],
            [_, _, _, _, _, _, _, _, X, _],
            [_, X, _, X, _, _, X, _, _, _],
            [_, _, _, X, _, X, _, X, _, X],
            [_, X, _, _, X, _, _, _, X, _],
            [_, _, X, _, _, _, _, _, _, X]
        ])
        expected_solution_str = (
            ' ┌───────────┐  ┌─────┐  ┌──┐ \n'
            ' │  ┌─────┐  │  │  ┌──┘  │  │ \n'
            ' │  │  ┌──┘  │  │  └─────┘  │ \n'
            ' └──┘  │  ┌──┘  └────────┐  │ \n'
            ' ┌──┐  │  │  ┌───────────┘  │ \n'
            ' │  │  │  │  └───────────┐  │ \n'
            ' │  └──┘  └─────┐  ┌─────┘  │ \n'
            ' │  ┌─────┐  ┌──┘  └──┐  ┌──┘ \n'
            ' │  └──┐  │  └────────┘  └──┐ \n'
            ' └─────┘  └─────────────────┘ '
        )

        game_solver = EverySecondTurnSolver(grid)
        solution = game_solver.get_solution()
        self.assertEqual(expected_solution_str, str(solution))
        other_solution = game_solver.get_other_solution()
        self.assertEqual(IslandGrid.empty(), other_solution)

    def test_solution_16_16_0zny2(self):
        """https://gridpuzzle.com/every-second-turn/0zny2"""

        grid = Grid([
            [X, _, X, _, X, _, X, _, _, _, _, _, _, _, X, _],
            [_, _, _, _, _, X, _, _, X, _, _, _, X, _, _, _],
            [_, _, _, X, _, _, _, X, _, X, _, _, _, X, _, _],
            [_, _, _, X, _, X, X, _, _, _, _, X, _, _, _, X],
            [_, X, _, _, X, _, _, _, X, _, X, _, X, _, X, _],
            [_, _, _, _, X, _, _, _, _, X, _, X, _, X, _, _],
            [X, _, _, _, _, X, _, _, _, _, _, X, _, _, X, _],
            [_, X, _, X, X, _, _, X, _, X, X, _, X, _, _, X],
            [X, _, X, _, _, X, _, _, _, X, _, _, X, X, _, _],
            [_, _, X, _, _, X, _, X, _, _, _, X, _, _, _, X],
            [X, _, _, _, X, _, _, _, X, _, _, X, _, X, _, _],
            [_, X, _, _, _, _, X, _, X, _, X, _, _, _, X, _],
            [_, _, X, X, _, _, _, X, _, _, X, _, _, X, _, _],
            [X, _, _, _, _, X, X, _, _, X, _, _, X, _, _, X],
            [_, X, _, _, X, _, _, X, X, _, _, X, _, _, _, _],
            [_, _, _, X, _, _, X, _, _, X, _, _, _, _, _, X]
        ])
        expected_solution_str = (
            ' ┌──┐  ┌──┐  ┌──┐  ┌────────────────────┐  ┌──┐ \n'
            ' │  │  │  │  │  └──┘  ┌──┐  ┌────────┐  │  │  │ \n'
            ' │  │  │  └──┘  ┌─────┘  │  └─────┐  │  └──┘  │ \n'
            ' │  │  └──┐  ┌──┘  ┌─────┘  ┌─────┘  └────────┘ \n'
            ' │  └─────┘  └──┐  │  ┌──┐  │  ┌──┐  ┌──┐  ┌──┐ \n'
            ' └───────────┐  │  │  │  │  └──┘  └──┘  └──┘  │ \n'
            ' ┌────────┐  └──┘  │  │  │  ┌─────┐  ┌─────┐  │ \n'
            ' └──┐  ┌──┘  ┌──┐  └──┘  └──┘  ┌──┘  └──┐  └──┘ \n'
            ' ┌──┘  └──┐  │  └────────┐  ┌──┘  ┌──┐  └─────┐ \n'
            ' └─────┐  │  └──┐  ┌──┐  │  │  ┌──┘  │  ┌─────┘ \n'
            ' ┌──┐  │  │  ┌──┘  │  │  └──┘  │  ┌──┘  └─────┐ \n'
            ' │  └──┘  │  └─────┘  └──┐  ┌──┘  │  ┌─────┐  │ \n'
            ' └─────┐  └────────┐  ┌──┘  │  ┌──┘  │  ┌──┘  │ \n'
            ' ┌──┐  │  ┌─────┐  └──┘  ┌──┘  │  ┌──┘  └─────┘ \n'
            ' │  └──┘  │  ┌──┘  ┌──┐  └──┐  │  └───────────┐ \n'
            ' └────────┘  └─────┘  └─────┘  └──────────────┘ '
        )

        game_solver = EverySecondTurnSolver(grid)
        solution = game_solver.get_solution()
        self.assertEqual(expected_solution_str, str(solution))
        other_solution = game_solver.get_other_solution()
        self.assertEqual(IslandGrid.empty(), other_solution)


if __name__ == '__main__':
    unittest.main()

File: DotNet/EverySecondTurnSolver.Tests/EverySecondTurnSolverTest.cs
using Xunit;

namespace EverySecondTurnSolver.Tests;

public class EverySecondTurnSolverTest
{
    private const char X = EverySecondTurnSolver.Circle;
    private const char _ = EverySecondTurnSolver.Empty;

    [Fact]
    public void TestSolution_6_6_37g81()
    {
        // https://gridpuzzle.com/every-second-turn/37g81
        var gridData = new char[,]
        {
            { _, _, _, _, _, X },
            { _, X, _, _, _, _ },
            { _, _, _, _, X, _ },
            { X, _, _, X, _, _ },
            { _, _, X, _, _, X },
            { X, _, _, _, _, _ },
        };

        const string expectedSolutionStr =
            " ┌──────────────┐ \n" +
            " │  ┌────────┐  │ \n" +
            " │  │  ┌─────┘  │ \n" +
            " └──┘  │  ┌─────┘ \n" +
            " ┌─────┘  └─────┐ \n" +
            " └──────────────┘ ";

        var solver = new EverySecondTurnSolver(gridData);
        var solution = solver.GetSolution();
        Assert.NotNull(solution);
        Assert.Equal(expectedSolutionStr, solution.ToString());

        var otherSolution = solver.GetOtherSolution();
        Assert.True(otherSolution == null || otherSolution.Rows == 0);
    }

    [Fact]
    public void TestSolution_6_6_3e9k2()
    {
        // https://gridpuzzle.com/every-second-turn/3e9k2
        var gridData = new char[,]
        {
            { X, _, _, _, _, _ },
            { _, _, _, _, _, X },
            { _, _, X, _, _, _ },
            { _, _, _, _, X, _ },
            { _, X, _, _, _, _ },
            { _, _, _, _, _, X }
        };

        const string expectedSolutionStr =
            " ┌──────────────┐ \n" +
            " │  ┌───────────┘ \n" +
            " │  │  ┌────────┐ \n" +
            " │  │  └─────┐  │ \n" +
            " │  └────────┘  │ \n" +
            " └──────────────┘ ";

        var solver = new EverySecondTurnSolver(gridData);
        var solution = solver.GetSolution();
        Assert.NotNull(solution);
        Assert.Equal(expectedSolutionStr, solution.ToString());

        var otherSolution = solver.GetOtherSolution();
        Assert.True(otherSolution == null || otherSolution.Rows == 0);
    }

    [Fact]
    public void TestSolution_10_10_0ygm8()
    {
        // https://gridpuzzle.com/every-second-turn/0ygm8
        var gridData = new[,]
        {
            { X, _, _, _, _, X, _, _, X, _ },
            { _, _, _, X, _, _, _, X, _, _ },
            { _, _, X, _, _, _, X, _, _, _ },
            { _, X, _, _, X, _, _, _, X, _ },
            { X, _, _, _, X, _, _, _, _, _ },
            { _, _, _, _, _, _, _, _, X, _ },
            { _, X, _, X, _, _, X, _, _, _ },
            { _, _, _, X, _, X, _, X, _, X },
            { _, X, _, _, X, _, _, _, X, _ },
            { _, _, X, _, _, _, _, _, _, X }
        };

        const string expectedSolutionStr =
            " ┌───────────┐  ┌─────┐  ┌──┐ \n" +
            " │  ┌─────┐  │  │  ┌──┘  │  │ \n" +
            " │  │  ┌──┘  │  │  └─────┘  │ \n" +
            " └──┘  │  ┌──┘  └────────┐  │ \n" +
            " ┌──┐  │  │  ┌───────────┘  │ \n" +
            " │  │  │  │  └───────────┐  │ \n" +
            " │  └──┘  └─────┐  ┌─────┘  │ \n" +
            " │  ┌─────┐  ┌──┘  └──┐  ┌──┘ \n" +
            " │  └──┐  │  └────────┘  └──┐ \n" +
            " └─────┘  └─────────────────┘ ";

        var solver = new EverySecondTurnSolver(gridData);
        var solution = solver.GetSolution();
        Assert.NotNull(solution);
        Assert.Equal(expectedSolutionStr, solution.ToString());

        var otherSolution = solver.GetOtherSolution();
        Assert.True(otherSolution == null || otherSolution.Rows == 0);
    }

    [Fact]
    public void TestSolution_16_16_0zny2()
    {
        // https://gridpuzzle.com/every-second-turn/0zny2
        var gridData = new[,]
        {
            { X, _, X, _, X, _, X, _, _, _, _, _, _, _, X, _ },
            { _, _, _, _, _, X, _, _, X, _, _, _, X, _, _, _ },
            { _, _, _, X, _, _, _, X, _, X, _, _, _, X, _, _ },
            { _, _, _, X, _, X, X, _, _, _, _, X, _, _, _, X },
            { _, X, _, _, X, _, _, _, X, _, X, _, X, _, X, _ },
            { _, _, _, _, X, _, _, _, _, X, _, X, _, X, _, _ },
            { X, _, _, _, _, X, _, _, _, _, _, X, _, _, X, _ },
            { _, X, _, X, X, _, _, X, _, X, X, _, X, _, _, X },
            { X, _, X, _, _, X, _, _, _, X, _, _, X, X, _, _ },
            { _, _, X, _, _, X, _, X, _, _, _, X, _, _, _, X },
            { X, _, _, _, X, _, _, _, X, _, _, X, _, X, _, _ },
            { _, X, _, _, _, _, X, _, X, _, X, _, _, _, X, _ },
            { _, _, X, X, _, _, _, X, _, _, X, _, _, X, _, _ },
            { X, _, _, _, _, X, X, _, _, X, _, _, X, _, _, X },
            { _, X, _, _, X, _, _, X, X, _, _, X, _, _, _, _ },
            { _, _, _, X, _, _, X, _, _, X, _, _, _, _, _, X }
        };

        const string expectedSolutionStr =
            " ┌──┐  ┌──┐  ┌──┐  ┌────────────────────┐  ┌──┐ \n" +
            " │  │  │  │  │  └──┘  ┌──┐  ┌────────┐  │  │  │ \n" +
            " │  │  │  └──┘  ┌─────┘  │  └─────┐  │  └──┘  │ \n" +
            " │  │  └──┐  ┌──┘  ┌─────┘  ┌─────┘  └────────┘ \n" +
            " │  └─────┘  └──┐  │  ┌──┐  │  ┌──┐  ┌──┐  ┌──┐ \n" +
            " └───────────┐  │  │  │  │  └──┘  └──┘  └──┘  │ \n" +
            " ┌────────┐  └──┘  │  │  │  ┌─────┐  ┌─────┐  │ \n" +
            " └──┐  ┌──┘  ┌──┐  └──┘  └──┘  ┌──┘  └──┐  └──┘ \n" +
            " ┌──┘  └──┐  │  └────────┐  ┌──┘  ┌──┐  └─────┐ \n" +
            " └─────┐  │  └──┐  ┌──┐  │  │  ┌──┘  │  ┌─────┘ \n" +
            " ┌──┐  │  │  ┌──┘  │  │  └──┘  │  ┌──┘  └─────┐ \n" +
            " │  └──┘  │  └─────┘  └──┐  ┌──┘  │  ┌─────┐  │ \n" +
            " └─────┐  └────────┐  ┌──┘  │  ┌──┘  │  ┌──┘  │ \n" +
            " ┌──┐  │  ┌─────┐  └──┘  ┌──┘  │  ┌──┘  └─────┘ \n" +
            " │  └──┘  │  ┌──┘  ┌──┐  └──┐  │  └───────────┐ \n" +
            " └────────┘  └─────┘  └─────┘  └──────────────┘ ";

        var solver = new EverySecondTurnSolver(gridData);
        var solution = solver.GetSolution();
        Assert.NotNull(solution);
        Assert.Equal(expectedSolutionStr, solution.ToString());

        var otherSolution = solver.GetOtherSolution();
        Assert.True(otherSolution == null || otherSolution.Rows == 0);
    }
}